lib/spear/connection.ex

```diff
@@ -115,22 +116,94 @@ defmodule Spear.Connection do
   defp process_response(_unknown, state), do: state
 
   defp stream_body(state, request_ref, messages) do
-    Enum.reduce_while(messages, {:ok, state}, &stream_body_message(&1, &2, request_ref))
+    Enum.reduce_while(
+      messages,
+      {:ok, state},
+      &stream_body_message(&1, &2, request_ref)
+    )
   end
 
   defp stream_body_message(message, {:ok, state}, request_ref) do
-    {wire_data, _byte_size} = Request.to_wire_data(message)
-
-    stream_result =
-      Mint.HTTP.stream_request_body(
-        state.conn,
-        request_ref,
-        wire_data
-      )
-
-    case stream_result do
-      {:ok, conn} -> {:cont, {:ok, put_in(state.conn, conn)}}
-      {:error, conn, reason} -> {:halt, {:error, put_in(state.conn, conn), reason}}
+    {wire_data, byte_size} = Request.to_wire_data(message)
+    smallest_window = get_smallest_window(state.conn, request_ref)
+
+    with false <- byte_size > smallest_window,
+         {:ok, conn} <- Mint.HTTP.stream_request_body(state.conn, request_ref, wire_data) do
+      {:cont, {:ok, put_in(state.conn, conn)}}
+    else
+      _window_too_small? = true ->
+        recv_until_window_size_increase(state, smallest_window, request_ref)
+
+      {:error, conn, reason} ->
+        {:halt, {:error, put_in(state.conn, conn), reason}}
+    end
+  end
+
+  defp recv_until_window_size_increase(state, current_window, request_ref) do
+    params = {current_window, request_ref}
+
+    {:ok, state}
+    |> do_stage(:recv_responses, params)
+    |> do_stage(:check_window_size, params)
+    |> emit_stage_results()
+  end
+
+  defp do_stage({:error, state, reason}, _stage, _params), do: {:error, state, reason}
+
+  defp do_stage({:ok, state}, :recv_responses, {_current_window, _request_ref}) do
+    case receive_next_and_stream(state.conn) do
+      {:ok, conn, responses} ->
+        state =
+          put_in(state.conn, conn)
+          |> handle_responses(responses)
+
+        {:ok, state}
+
+      {:error, conn, reason} ->
+        state = put_in(state.conn, conn)
+
+        {:error, state, reason}
+    end
+  end
+
+  defp do_stage({:ok, state}, :check_window_size, {current_window, request_ref}) do
+    new_window = get_smallest_window(state.conn, request_ref)
+
+    if new_window > current_window do
+      {:ok, state}
+    else
+      # if the window has not gotten bigger, loop back to the prior stage
+      # in the pipeline: block and wait for more messages from the server
+      do_stage({:ok, state}, :recv_responses, {new_window, request_ref})
+    end
+  end
+
+  defp emit_stage_results({:ok, state}), do: {:cont, {:ok, state}}
+  defp emit_stage_results({:error, state, reason}), do: {:halt, {:error, state, reason}}
+
+  defp receive_next_and_stream(conn) do
+    # YARD allow customization of timeout?
+    receive do
+      message when Mint.HTTP.is_connection_message(conn, message) ->
```

Comment:
this macro is why the `mix.exs` has a bump to requiring `~> 1.10`: this macro depends on some of the guard improvements in 1.10.0

need to use this here so we only receive Mint.HTTP messages and not accidentally intercept genserver calls/casts